Sources/MapboxMaps/Annotations/ViewAnnotation.swift

```diff
@@ -0,0 +1,206 @@
+
+    public func getViewAnnotationOptions(byAnnotationView view: AnnotationView) -> ViewAnnotationOptions? {
+        guard let id = annotationIdByViews[view] else {
+            // TODO: gracefully handle error
+            return nil
+        }
+        return try? mapboxMap.getViewAnnotationOptions(forId: id)
+    }
+
+    func getViewAnnotationOptions(byFeatureId identifier: String) -> ViewAnnotationOptions? {
+        return getViewAnnotation(byFeatureId: identifier).map({ view in getViewAnnotationOptions(byAnnotationView: view) }) ?? nil
+    }
+
+    // MARK: - Internal functions
+
+    // Iterate through and update all view annotations
+    // First update the position of the views based on the placement info from GL-Native
+    // Then hide the views which are off screen
+    internal func placeAnnotations(positions: [ViewAnnotationPositionDescriptor]) {
+        var visibleAnnotationIds: Set<String> = []
+
+        for position in positions {
+            validateAnnotation(byAnnotationId: position.identifier)
+            guard let annotationView = annotationViewsById[position.identifier] else {
+                fatalError()
+            }
+            // TODO: Check if position depends on the device's pixel ratio. (In a previous commit this was divided by two for some reason.)
+            annotationView.frame = CGRect(
+                origin: position.leftTopCoordinate.point,
+                size: CGSize(width: CGFloat(position.width), height: CGFloat(position.height))
+            )
+
+            annotationView.setVisibilityWithoutUpdate(isHidden: false)
+            visibleAnnotationIds.insert(position.identifier)
+        }
+
+        let annotationsToHide = Set<String>(annotationViewsById.keys).subtracting(visibleAnnotationIds)
+        for id in annotationsToHide {
+            validateAnnotation(byAnnotationId: id)
+            if let annotationView = annotationViewsById[id] {
+                annotationView.setVisibilityWithoutUpdate(isHidden: true)
+            }
+        }
+    }
+
+    internal func validateAnnotation(byAnnotationId id: String) {
+        guard let annotationView = annotationViewsById[id] else { return }
+        // If the user explicitly removed the ViewAnnotation or it's warpped view
+        // then we need to remove it from our layout calculation
+        if annotationView.subviews.isEmpty || annotationView.superview == nil {
+            _ = removeViewAnnotation(annotationView)
+        }
+        if let wrappedView = annotationView.subviews.first, wrappedView.isHidden {
+            // View is still considered for layout calculation, users should not modify the visibility of the wrapped view
+            // TODO: Print warning
+        }
+    }
+
+}
+
+extension ViewAnnotationManagerImpl: ViewAnnotationPositionsListener {
+
+    public func onViewAnnotationPositionsUpdate(forPositions positions: [ViewAnnotationPositionDescriptor]) {
+        placeAnnotations(positions: positions)
+    }
+
+}
```

Comment:
When we make `ViewAnnotationManagerImpl` public, let's avoid making this method public by following the delegation pattern used in a few other places like `DelegatingMapClient`

Reply:
Thanks for the tip, I've added an implementation based on that. Although the name of the class might be a bit weird: https://github.com/mapbox/mapbox-maps-ios/pull/776/files#diff-6be4c15b3e0af24102f7ee6ab1a8187765ac65a9c83b8f314070d6eec3be7791R8

Reply:
Name seems good to me